Add Substrait-to-Velox conversion

rui-mo opened the pull request:
This pr is targeted to add the base layout for Substrait-to-Velox conversion. With this change, the base layout of Substrait-to-Velox conversion would be decided. Some of Substrait plan representations can be converted into Velox plan for computing. More conversions support will be added in following pull requests. 

Based on [Substrait Commit](https://github.com/substrait-io/substrait/commit/9d9805be19c1d606dc2811395f203857db782872).

The used function names are based on: [Substrait PR](https://github.com/substrait-io/substrait/pull/147).

pedroerp commented:
This is great, thanks for contributing this back to Velox! Before I start looking at the code, I wonder if you already had thoughts on how we would add the dependent Substrait headers to ensure these files build on circleCI and that are easily usable by Velox developers. 

I happy to help brainstorming if you haven't!

rui-mo commented:
> This is great, thanks for contributing this back to Velox! Before I start looking at the code, I wonder if you already had thoughts on how we would add the dependent Substrait headers to ensure these files build on circleCI and that are easily usable by Velox developers.
> 
> I happy to help brainstorming if you haven't!

Thanks for you advice. The Substrait hearders can be automatically generated with some commands in CMakeLists. We will soon try to port this part from our implementations to Velox, thanks!

jacques-n commented:
One of the things we need to do in Substrait is to start releasing releases so people don't have to depend on git hash versions. @cpcloud, any chance you could take a look at https://github.com/substrait-io/substrait/issues/141 to help with this?

rui-mo commented:
Substrait headers can be automatically generated, and the code compilation works now. A unit test was added, in which input would be a dumped Substrait plan, and output would be Velox computing result. Working on solving the multiple reference issues for the compilation of unit test. For now, Substrait proto files were put into Velox, if you have other opinion, please let me know, thanks.

winningsix commented:
@pedroerp  @jacques-n Quick question: Will this PR depend on Substrait release or not? Your thoughts here. I will suggest to live with what Substrait already has. And rebase on release in a follow-up PR.

jacques-n commented:
My  recommendation would be that if the team wants to merge this before a release, we at least ensure it is from a commit in Substrait main branch. I don't think that it should super long to get to a first release. Thoughts @cpcloud?

winningsix commented:
@jacques-n which commmit should be based on?

jacques-n commented:
Something in Substrait main branch. Preferably latest available commit.

rui-mo commented:
OK, will rebase this PR to the latest available Substrait commit before getting the release.

rui-mo commented:
Updated Substrait to the latest available commit and added a unit test to read Json file and test Velox's computing. All Github checks were passed. This PR is ready for review now, thanks.

pedroerp reviewed (commented):
Thanks for looking into this! A few comments inline:

rui-mo commented on velox/substrait_converter/SubstraitToVeloxExpr.cpp:
```diff
@@ -0,0 +1,301 @@
+  }
+}
+
+std::shared_ptr<const core::ITypedExpr>
```
The return type is not limited to CallTypedExpr, but can also be other expressions extend ITypedExpr, like FieldAccessTypedExpr, ConstantTypedExpr etc.
Reference: https://github.com/facebookincubator/velox/blob/main/velox/core/Expressions.h

rui-mo commented on velox/CMakeLists.txt:
```diff
@@ -64,3 +64,8 @@ endif()
+
+# substrait converter
+if(${VELOX_ENABLE_SUBSTRAIT})
+  add_subdirectory(substrait_converter)
```
Renamed in the latest [commit](https://github.com/facebookincubator/velox/pull/1048/commits/661ed968a56078eccc5d0099a5ac69db97e76cb9).

rui-mo commented on velox/substrait_converter/SubstraitToVeloxExpr.h:
```diff
@@ -0,0 +1,83 @@
+class SubstraitVeloxExprConverter {
+ public:
+  SubstraitVeloxExprConverter(
+      const std::shared_ptr<SubstraitParser>& sub_parser,
```
Fixed in the latest [commit](https://github.com/facebookincubator/velox/pull/1048/commits/661ed968a56078eccc5d0099a5ac69db97e76cb9).

rui-mo commented on velox/substrait_converter/SubstraitToVeloxExpr.h:
```diff
@@ -0,0 +1,83 @@
+  std::shared_ptr<const core::FieldAccessTypedExpr> toVeloxExpr(
+      const substrait::Expression::FieldReference& sfield,
+      const int32_t& input_plan_node_id);
+  std::shared_ptr<const core::ITypedExpr> toVeloxExpr(
```
Fixed.

rui-mo commented on velox/substrait_converter/SubstraitToVeloxExpr.h:
```diff
@@ -0,0 +1,83 @@
+// This class is used to convert Substrait representations to Velox expressions.
+class SubstraitVeloxExprConverter {
+ public:
+  SubstraitVeloxExprConverter(
```
Comments were added.

rui-mo commented on velox/substrait_converter/SubstraitToVeloxExpr.h:
```diff
@@ -0,0 +1,83 @@
+      const substrait::Expression& sfilter);
+
+ private:
+  void getFlatConditions(
```
Comments were added.

rui-mo commented on velox/substrait_converter/SubstraitToVeloxExpr.h:
```diff
@@ -0,0 +1,83 @@
+#include "velox/functions/prestosql/aggregates/SumAggregate.h"
+#include "velox/functions/prestosql/registration/RegistrationFunctions.h"
+#include "velox/type/Filter.h"
+#include "velox/type/Subfield.h"
```
Yes, the unneeded header were removed from this header. The implementations were in the .cpp file.

rui-mo commented on velox/substrait_converter/SubstraitToVeloxExpr.h:
```diff
@@ -0,0 +1,83 @@
+      const substrait::Expression& sfilter,
+      std::vector<substrait::Expression_ScalarFunction>* scalar_functions);
+  std::shared_ptr<SubstraitParser> sub_parser_;
+  std::unordered_map<uint64_t, std::string> functions_map_;
```
Fixed.

rui-mo commented on velox/substrait_converter/SubstraitToVeloxExpr.h:
```diff
@@ -0,0 +1,83 @@
+
+#pragma once
+
+#include <folly/executors/IOThreadPoolExecutor.h>
```
Removed.

rui-mo commented on velox/substrait_converter/SubstraitToVeloxExpr.cpp:
```diff
@@ -0,0 +1,301 @@
+SubstraitVeloxExprConverter::SubstraitVeloxExprConverter(
+    const std::shared_ptr<SubstraitParser>& sub_parser,
+    const std::unordered_map<uint64_t, std::string>& functions_map) {
+  sub_parser_ = sub_parser;
```
Fixed.

rui-mo commented on velox/substrait_converter/SubstraitToVeloxExpr.cpp:
```diff
@@ -0,0 +1,301 @@
+std::shared_ptr<const core::FieldAccessTypedExpr>
+SubstraitVeloxExprConverter::toVeloxExpr(
+    const substrait::Expression::FieldReference& sfield,
+    const int32_t& input_plan_node_id) {
```
Fixed.

rui-mo commented on velox/substrait_converter/SubstraitToVeloxExpr.cpp:
```diff
@@ -0,0 +1,301 @@
+        kDirectReference: {
+      auto dref = sfield.direct_reference();
+      int32_t col_idx = parseReferenceSegment(dref);
+      auto field_name = sub_parser_->makeNodeName(input_plan_node_id, col_idx);
```
Fixed.

rui-mo commented on velox/substrait_converter/SubstraitToVeloxExpr.cpp:
```diff
@@ -0,0 +1,301 @@
+      auto dref = sfield.direct_reference();
+      int32_t col_idx = parseReferenceSegment(dref);
+      auto field_name = sub_parser_->makeNodeName(input_plan_node_id, col_idx);
+      // FIXME: get the input type
```
Fixed in the latest [commit](https://github.com/facebookincubator/velox/pull/1048/commits/661ed968a56078eccc5d0099a5ac69db97e76cb9).

rui-mo commented on velox/substrait_converter/SubstraitToVeloxExpr.cpp:
```diff
@@ -0,0 +1,301 @@
+      auto dref = sfield.direct_reference();
+      int32_t col_idx = parseReferenceSegment(dref);
+      auto field_name = sub_parser_->makeNodeName(input_plan_node_id, col_idx);
+      // FIXME: get the input type
```
Fixed.

rui-mo commented on velox/substrait_converter/SubstraitToVeloxExpr.cpp:
```diff
@@ -0,0 +1,301 @@
+    }
+    case substrait::Expression::FieldReference::ReferenceTypeCase::
+        kMaskedReference: {
+      throw new std::runtime_error("not supported");
```
Fixed. VELOX_NYI() and VELOX_FAIL() are used now instead of throwing the runtime error.

rui-mo commented on velox/substrait_converter/SubstraitToVeloxExpr.cpp:
```diff
@@ -0,0 +1,301 @@
+          DOUBLE(), field_name);
+      break;
+    }
+    case substrait::Expression::FieldReference::ReferenceTypeCase::
```
Fixed.

rui-mo commented on velox/substrait_converter/SubstraitToVeloxExpr.cpp:
```diff
@@ -0,0 +1,301 @@
+    const substrait::Expression::ScalarFunction& sfunc,
+    const int32_t& input_plan_node_id) {
+  std::vector<std::shared_ptr<const core::ITypedExpr>> params;
+  for (auto& sarg : sfunc.args()) {
```
Fixed all the cases like this.

rui-mo commented on velox/substrait_converter/SubstraitToVeloxExpr.cpp:
```diff
@@ -0,0 +1,301 @@
+    auto expr = toVeloxExpr(sarg, input_plan_node_id);
+    params.push_back(expr);
+  }
+  auto function_id = sfunc.function_reference();
```
Fixed.

rui-mo commented on velox/substrait_converter/SubstraitToVeloxExpr.cpp:
```diff
@@ -0,0 +1,301 @@
+SubstraitVeloxExprConverter::toVeloxExpr(
+    const substrait::Expression::Literal& slit) {
+  switch (slit.literal_type_case()) {
+    case substrait::Expression_Literal::LiteralTypeCase::kFp64: {
```
Fixed.

rui-mo commented on velox/substrait_converter/SubstraitToVeloxExpr.cpp:
```diff
@@ -0,0 +1,301 @@
+      return std::make_shared<core::ConstantTypedExpr>(val);
+      break;
+    }
+    case substrait::Expression_Literal::LiteralTypeCase::kBoolean: {
```
Fixed.

pedroerp commented on velox/substrait/SubstraitToVeloxExpr.h:
```diff
@@ -0,0 +1,80 @@
+
+  /// The Substrait parser used to convert Substrait representations into
+  /// recognizable representations.
+  std::shared_ptr<SubstraitParser> subParser_;
```
nit: blank line before the comment block

rui-mo replied:
Fixed the cases like this.

pedroerp commented on velox/substrait/SubstraitToVeloxExpr.h:
```diff
@@ -0,0 +1,80 @@
+  std::shared_ptr<SubstraitParser> subParser_;
+  /// The map storing the relations between the function id and the function
+  /// name.
+  std::unordered_map<uint64_t, std::string> functionMap_;
```
same here

rui-mo replied:
Fixed.

pedroerp commented on velox/substrait/SubstraitToVeloxExpr.cpp:
```diff
@@ -0,0 +1,267 @@
+
+void SubstraitVeloxExprConverter::getFlatConditions(
+    const ::substrait::Expression& sFilter,
+    std::vector<::substrait::Expression_ScalarFunction>* scalarFunctions) {
```
pointers are usually to specify this parameter is optional, but it doesn't seem to be the case. Could we take it as a non-const ref instead?

pedroerp commented on velox/substrait/SubstraitToVeloxExpr.cpp:
```diff
@@ -0,0 +1,267 @@
+      break;
+    }
+    default:
+      VELOX_NYI("GetFlatConditions not supported for type '{}'", typeCase);
```
Is there a `toString()` sort of method for typeCase? I'm assuming this is a enum, so we would end up printing the integer value, which is not very helpful to users.

rui-mo replied:
Yes, the typeCase is just integer value. But after searching out, did not find an elegant way to map this integer value to the key word in enum. I found two options not quite simple:

- Construct a map to store the mapping relations
- Use a tool to map the integer back to key word: https://github.com/Neargye/magic_enum

pedroerp commented on velox/substrait/SubstraitToVeloxExpr.cpp:
```diff
@@ -0,0 +1,267 @@
+SubstraitVeloxExprConverter::toVeloxExpr(
+    const ::substrait::Expression::ScalarFunction& sFunc,
+    const int32_t& inputPlanNodeId) {
+  std::vector<std::shared_ptr<const core::ITypedExpr>> params;
```
reserve it first?

`params.reserve(sFunc.args().size());`

rui-mo replied:
Fixed the cases like this.

pedroerp commented on velox/substrait/SubstraitToVeloxExpr.cpp:
```diff
@@ -0,0 +1,267 @@
+  std::vector<std::shared_ptr<const core::ITypedExpr>> params;
+  for (const auto& sArg : sFunc.args()) {
+    auto expr = toVeloxExpr(sArg, inputPlanNodeId);
+    params.push_back(expr);
```
nit: one-liner?
```
    params.emplace_back(toVeloxExpr(sArg, inputPlanNodeId));
```

pedroerp commented on velox/substrait/SubstraitToVeloxExpr.cpp:
```diff
@@ -0,0 +1,267 @@
+    const ::substrait::Expression::Literal& sLit) {
+  auto typeCase = sLit.literal_type_case();
+  switch (typeCase) {
+    case ::substrait::Expression_Literal::LiteralTypeCase::kFp64: {
```
nit: don't need the code block here '{}'

rui-mo replied:
Fixed the cases like this.

pedroerp commented on velox/substrait/SubstraitToVeloxExpr.cpp:
```diff
@@ -0,0 +1,267 @@
+  auto typeCase = sExpr.rex_type_case();
+  switch (typeCase) {
+    case ::substrait::Expression::RexTypeCase::kLiteral: {
+      veloxExpr = toVeloxExpr(sExpr.literal());
```
nit: return here directly?

rui-mo replied:
Fixed the cases like this.

pedroerp commented on velox/substrait/SubstraitToVeloxExpr.h:
```diff
@@ -0,0 +1,80 @@
+  /// name.
+  std::unordered_map<uint64_t, std::string> functionMap_;
+  /// This class contains the needed infos for Filter Pushdown.
+  class FilterInfo;
```
do we need to declare this class here? Seems like we could get away with just defining it as a free class in the .cpp file.

pedroerp commented on velox/substrait/SubstraitToVeloxExpr.cpp:
```diff
@@ -0,0 +1,267 @@
+// TODO: Support different types here.
+class SubstraitVeloxExprConverter::FilterInfo {
+ public:
+  FilterInfo() {}
```
nit: no need to explicitly specify it in this case

rui-mo replied:
Fixed the cases like this.

pedroerp commented on velox/substrait/SubstraitToVeloxExpr.h:
```diff
@@ -0,0 +1,80 @@
+  /// the relation key words, including AND, OR, and etc. Currently, only
+  /// AND is supported. This function is used to extract all the Substrait
+  /// conditions in the binary tree structure into a vector.
+  void getFlatConditions(
```
nit: maybe `flattenConditions()` could be more descriptive

rui-mo replied:
Fixed.

pedroerp commented on velox/substrait/SubstraitToVeloxExpr.h:
```diff
@@ -0,0 +1,80 @@
+      const int32_t& inputPlanNodeId);
+
+  /// Used to convert Substrait Filter into Velox SubfieldFilters.
+  connector::hive::SubfieldFilters toVeloxFilter(
```
I'm a bit confused not. If this class only converts expressions, shouldn't filter conversion be handled by the plan conversion api in the other file?

rui-mo replied:
Yes, the filter conversion will be used in TableScan, so it is actually related to the plan conversion. Moved this part into the plan converter.

pedroerp commented:
Thanks for addressing the comments! Just made a few more smaller comments here and there, but my main question is about where we place the filter conversion code.

jacques-n commented on velox/substrait/SubstraitToVeloxPlan.h:
```diff
@@ -0,0 +1,134 @@
+
+  /// The map storing the relations between the function id and the function
+  /// name. Will be constructed based on the Substrait representation.
+  std::unordered_map<uint64_t, std::string> functionMap_;
```
This pattern of a function map doesn't seem to correctly represent the variability possible in a Substrait plan. How is this map constructed? It should be done by reading specific other parts of the plan.

rui-mo replied:
Yes, I think we need to follow your suggestion to construct the function name into below format:
<function name>:<short_arg_type0>_<short_arg_type1>_..._<short_arg_typeN>
And on your opinion, when should we validate those functions based on yaml specifications? In Spark or in Velox? I think we need to validate them before starting the computing.

rui-mo commented:
Pushed one commit to fix the issues mentioned above. For the next, we will change the used function key words to be aligned with the specifications in Substrait yaml files.
The related function names are: "AND", "SUM", "MULTIPLY", "LESS_THAN", "LESS_THAN_OR_EQUAL",, "GREATER_THAN_OR_EQUAL", "IS_NOT_NULL".
For those Substrait does not have yet, we will raise a PR to Substrait to add them.

pedroerp commented on velox/substrait/tests/PlanConversionTest.cpp:
```diff
@@ -0,0 +1,510 @@
+  }
+};
+
+// This test will firstly generate mock TPC-H lineitem ORC file. Then, Velox's
```
Overall, this test suite has quite a lot of dependencies. Would it be possible to have a leaner test suite with multiple smaller cases exercising only the plan and expr conversion code? 

My point being that we don't necessarily need to execute the full query here, but rather just test that given an input substrait plan/expr, it generates the expect Velox plan/expr. Verifying that Velox correctly executes a given Velox plan/expr is already covered by other test suites. 

We could also build e2e tests like this one, but my suggestion would be to add this along other tpch test execution code.

rui-mo replied:
Thanks for your insightful comments.
This is just our first PR. Later, my colleagues and I will submit a series of PRs to improve the UT and add more function support. This UT ensures that the subsequent PRs would not break the functionalities currently supported. We plan to add the UTs you mentioned in our following PRs.

pedroerp commented on velox/substrait/SubstraitUtils.h:
```diff
@@ -0,0 +1,84 @@
+
+namespace facebook::velox::substrait {
+
+/// This class contains some common funcitons used to parse Substrait
```
typo

rui-mo replied:
Fixed.

pedroerp commented on velox/substrait/SubstraitUtils.h:
```diff
@@ -0,0 +1,84 @@
+  struct SubstraitType {
+    std::string type;
+    bool nullable;
+    SubstraitType(const std::string& subType, const bool& subNullable) {
```
use constructor intializers? Also, you might not need to explicitly specify this constructor - it's available by default (`SubstraitType{a, b} type`).

rui-mo replied:
Fixed.

pedroerp commented on velox/substrait/SubstraitUtils.h:
```diff
@@ -0,0 +1,84 @@
+  /// from a pre-constructed function map.
+  std::string findSubstraitFunction(
+      const std::unordered_map<uint64_t, std::string>& functionMap,
+      const uint64_t& id) const;
```
you can just pass primitive types by value.

rui-mo replied:
Fixed the cases like this.

pedroerp commented on velox/substrait/SubstraitUtils.h:
```diff
@@ -0,0 +1,84 @@
+  /// from a pre-constructed function map.
+  std::string findVeloxFunction(
+      const std::unordered_map<uint64_t, std::string>& functionMap,
+      const uint64_t& id) const;
```
same here

rui-mo replied:
Fixed.

pedroerp commented on velox/substrait/SubstraitUtils.h:
```diff
@@ -0,0 +1,84 @@
+  /// words. Key: the Substrait function key word, Value: the Velox function key
+  /// word.
+  const std::unordered_map<std::string, std::string> substraitVeloxFunctionMap =
+      {{"MULTIPLY", "multiply"}, {"SUM", "sum"}};
```
will we need to add a mapping to every function available in Velox here at some point?

rui-mo replied:
For those functions with different names in Substrait and Velox, a mapping relation is needed. Since we change to use the unified function names with Substrait, the two mapping relations here are not needed any more. Because in Substrait, they are also specified as "multiply" and "sum".

pedroerp commented on velox/substrait/SubstraitUtils.cpp:
```diff
@@ -0,0 +1,160 @@
+std::vector<std::shared_ptr<SubstraitParser::SubstraitType>>
+SubstraitParser::parseNamedStruct(const ::substrait::NamedStruct& namedStruct) {
+  const auto& sNames = namedStruct.names();
+  std::vector<std::string> nameList;
```
why do we need this vector?

rui-mo replied:
Removed.

pedroerp commented on velox/substrait/SubstraitUtils.cpp:
```diff
@@ -0,0 +1,160 @@
+  switch (typeCase) {
+    case ::substrait::Expression::ReferenceSegment::ReferenceTypeCase::
+        kStructField: {
+      auto sField = sRef.struct_field();
```
nit: one-liner
`
return sRef.struct_field().field();
`

rui-mo replied:
Fixed.

pedroerp commented on velox/substrait/SubstraitUtils.cpp:
```diff
@@ -0,0 +1,160 @@
+  if (functionMap.find(id) == functionMap.end()) {
+    VELOX_FAIL("Could not find function id {} in function map.", id);
+  }
+  std::unordered_map<uint64_t, std::string>& map =
```
do you actually want to add the key if it doesn't exist in here? Look like you should call .find() on the map, then return a default value (or throw) if it doesn't exist?

rui-mo replied:
Thanks for your comment.
The function map is pre-constructed in the Substrait plan. If the key doesn't exist, then the function map should be wrongly constructed or the key is invalid. Therefore, an exception is thrown in this case. The function map is decided on the Substrait plan producer side, so the unfounded keys should not be added here.

pedroerp commented on velox/substrait/SubstraitUtils.cpp:
```diff
@@ -0,0 +1,160 @@
+    VELOX_FAIL(
+        "Could not find Substrait function {} in function map.", subFunc);
+  }
+  std::unordered_map<std::string, std::string>& map =
```
same here.

rui-mo replied:
When the function has different names in Substrait and Velox, a mapping relation is needed.
Here we want to map the Substrait key word to Velox's. If not founded, their names are regarded to be the same, so the input should just be returned.

pedroerp commented on velox/substrait/TypeUtils.cpp:
```diff
@@ -0,0 +1,72 @@
+
+bool isString(const TypePtr& type) {
+  switch (type->kind()) {
+    case TypeKind::VARCHAR:
```
shouldn't VARBINARY be string type as well?

rui-mo replied:
Fixed.

pedroerp commented on velox/substrait/TypeUtils.cpp:
```diff
@@ -0,0 +1,72 @@
+
+namespace facebook::velox::substrait {
+
+bool isPrimitive(const TypePtr& type) {
```
can we just use the traits available in velox/type/Type.h here?

rui-mo replied:
Yes, we can use that. Removed this method.

pedroerp commented on velox/substrait/TypeUtils.cpp:
```diff
@@ -0,0 +1,72 @@
+  }
+}
+
+TypePtr toVeloxType(const std::string& typeName) {
```
I think other than this function, the functionality required should already be available in velox/type/Type.h

rui-mo replied:
Changed to use a method in Type.h called "mapNameToTypeKind" to replace this implementation.

pedroerp commented on velox/substrait/SubstraitToVeloxPlan.h:
```diff
@@ -0,0 +1,134 @@
+  std::shared_ptr<const core::PlanNode> toVeloxPlan(
+      const ::substrait::FilterRel& sFilter);
+
+  /// Used to convert Substrait ReadRel into Velox PlanNode.
```
can we explain what the parameters below mean?

rui-mo replied:
Added explanation for them.

pedroerp commented on velox/substrait/SubstraitToVeloxPlan.h:
```diff
@@ -0,0 +1,134 @@
+  }
+
+  /// Will add the paths of the files to be scanned into a vector.
+  void getPaths(std::vector<std::string>& paths) {
```
perhaps a more idiomatic way to achieve this is to have a function exposing the path_ vector as a const ref, and if you need to reuse this pattern inside the .cpp, this could be a free function inside an anonymous namespace on that file?

rui-mo replied:
Thanks for your advice.
Actually these methods are called outside the .cpp, and in this PR, they are called in the UT. Changed them to expose the vector as a const ref.

pedroerp commented on velox/substrait/SubstraitToVeloxPlan.cpp:
```diff
@@ -0,0 +1,504 @@
+    auto groupingExprs = grouping.grouping_expressions();
+    for (const auto& groupingExpr : groupingExprs) {
+      auto fieldExpr =
+          exprConverter_->toVeloxExpr(groupingExpr, inputPlanNodeId);
```
can we make the function return the specialized type directly (core::FieldAccessTypedExpr) so we don't need the cast below?

rui-mo replied:
Fixed in this way.

pedroerp commented on velox/substrait/SubstraitToVeloxPlan.cpp:
```diff
@@ -0,0 +1,504 @@
+  std::vector<std::string> projectOutNames;
+  std::vector<std::shared_ptr<const core::CallTypedExpr>> aggExprs;
+  aggExprs.reserve(sAgg.measures().size());
+  // Construct Velox Aggregate expressions.
```
nit: maybe use a blank line to separate code blocks?

rui-mo replied:
Fixed.

pedroerp commented on velox/substrait/SubstraitToVeloxPlan.cpp:
```diff
@@ -0,0 +1,504 @@
+std::shared_ptr<const core::PlanNode> SubstraitVeloxPlanConverter::toVeloxPlan(
+    const ::substrait::RelRoot& sRoot) {
+  const auto& sNames = sRoot.names();
+  int nameIdx = 0;
```
`int nameIdx = sNames.size();`

?

rui-mo replied:
Removed. It is not used currently.

pedroerp reviewed (commented):
Pheeew, getting there! A few more comments:

pedroerp reviewed (approved):
Thanks for iterating on this diff. There will likely be more improvements needed, but let's get this in and improve as we go.

Thanks for the contribution!

rui-mo commented:
Hi @pedroerp, thanks for your approval.
I note Facebook Internal - Linter reported 1 warning & 2 advices for my PR, but I cannot see the details due to lacking the permission. Do I need to fix them? If so, could you expose some details?

pedroerp commented:
> Hi @pedroerp, thanks for your approval. I note Facebook Internal - Linter reported 1 warning & 2 advices for my PR, but I cannot see the details due to lacking the permission. Do I need to fix them? If so, could you expose some details?

No worries, I'll have to add the internal building details, so I can take care of the linter warning. Thanks again for looking into this!
